autovalet_husky/scripts/find_parking_spot.py

```diff
@@ -0,0 +1,173 @@
+#!/usr/bin/env python
+'''
+Script to generate parking goal pose from detecting an april tag
+
+Date    : 6 Oct 2020
+
+Changelog:
+    subbu - refactoring to work with tag topic than TF
+    eva - generate 2 waypoints
+'''
+import numpy as np
+
+# Include ROS libs
+import rospy
+import tf2_ros
+from tf2_geometry_msgs import do_transform_pose
+from tf.transformations import euler_from_quaternion, quaternion_from_euler
+
+# Include messages
+from geometry_msgs.msg import PoseStamped, Quaternion, Pose
+from tf.transformations import quaternion_matrix, quaternion_from_matrix
+
+class parking_spot:
+    '''
+    Module for defining parking spot parameters
+    '''
+    def __init__(self, goal_topic, tag_topic, goal_frame_id, husky_frame_id, aruco_frame_name):
+        # for accumulating tag poses
+        self.count = 0
+        self.tfArray = []
+
+        # Setup listeners and talkers
+        self.pub = rospy.Publisher(goal_topic, PoseStamped, queue_size=1)
+        self.sub = rospy.Subscriber(tag_topic, PoseStamped, self.collect_tag_poses)
+
+        self.tf_buffer        = tf2_ros.Buffer(rospy.Duration(1200.0)) # Length of tf2 buffer (?)
+        self.tf_listener      = tf2_ros.TransformListener(self.tf_buffer)
+        self.aruco_frame_name = aruco_frame_name
+        self.goal_frame_id    = goal_frame_id
+        self.husky_frame_id   = husky_frame_id
+
+    def collect_tag_poses(self, tag_pose):
+        '''
+        Take in numOfTags tag poses and do RANSAC in the end to avoid outliers
+        '''
+        numOfTags = 20
+        if self.count < numOfTags: # collect tag poses
+            self.aruco_frame_id = tag_pose.header.frame_id
+
+            tf = self.tf_buffer.lookup_transform(self.goal_frame_id, # target_frame_id
+                                    self.aruco_frame_name, # source frame
+                                    rospy.Time(0), # get the tf at first available time
+                                    rospy.Duration(1.0)) # timeout after 1
+            self.tfArray.append(tf)
+            self.count += 1
+        elif self.count == numOfTags: # perform RANSAC on transformArray
+            self.tag_tf = self.tag_pose_RANSAC()
+            self.pub_two_goals()
+            self.count += 1
+
+    def pub_two_goals(self):
+        parking = False
+        pos1 = [0, 2, 1]
+        rot1 = [-np.pi/2, 0, -np.pi/2]
+        goal1 = self.generate_parking_goal(self.tag_tf, pos1, rot1)
+        while not parking:
+            rospy.sleep(0.5)
+            # keep publishing the first goal until the robot is close enough
+            if self.dist_to_first_goal(goal1) > 0.5:
+                self.pub.publish(goal1)
+                print("first goal published")
+            else:
+                pos2 = [0, 0, 4]
```

Comment:
maek a note saying this is a right(?) turn

Reply:
done